# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [8]:
import numpy as py
import pandas as pd

In [9]:
original_data = pd.read_csv("./data/e_commerce.csv")
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

#### 评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察植，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

In [12]:
original_data.sample(10)

InvoiceNo StockCode                        Description  Quantity  \
324951    565426     22562             MONSTERS STENCIL CRAFT         1   
229423   C557054    72807A  SET/3 ROSE CANDLE IN JEWELLED BOX        -2   
534102    581133     22422                TOOTHPASTE TUBE PEN        50   
126594    547088     82552                WASHROOM METAL SIGN        12   
263078    559933     84360   TABLE LAMP WHITE SHADE WOOD BASE         2   
30952     538884     22197               SMALL POPCORN HOLDER         5   
265656    560225     22752       SET 7 BABUSHKA NESTING BOXES         2   
440916    574550     21844                  RED RETROSPOT MUG         6   
52251     540698     22120      WELCOME  WOODEN BLOCK LETTERS        20   
232819    557330     22564             ALPHABET STENCIL CRAFT        12   

             InvoiceDate  UnitPrice  CustomerID         Country  
324951    9/4/2011 12:41       1.25     12851.0  United Kingdom  
229423   6/16/2011 14:43       4.25     17828.0           Malta  
534102   12/7/2011 12:55       0.12     14904.0  United Kingdom  
126594   3/20/2011 16:02       1.45     14223.0  United Kingdom  
263078    7/14/2011 9:54       5.95     16719.0  United Kingdom  
30952   12/14/2010 15:56       1.66         NaN  United Kingdom  
265656   7/15/2011 16:27       8.29         NaN  United Kingdom  
440916   11/4/2011 15:18       2.95     12484.0           Spain  
52251     1/11/2011 9:24      10.13     17450.0  United Kingdom  
232819   6/20/2011 10:20       1.25     14291.0  United Kingdom

###### 从抽样的10行数据来看，数据符合“每列是一个变量，每行是一个观察植，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

##### 数据干净程度评估

In [15]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


###### 从输出结果来看，数据共有541909条观察值，而Description、Customer ID变量存在缺失值。

此外，InvoiceData的数据类型应为日期，Customer ID的数据类型应为字符串，应当进行数据格式转换。

In [17]:
#评估缺失数据
#在了解Description存在缺失值后，根据条件提取出缺失观察值。
original_data[original_data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

In [18]:
#查看是否存在Description缺失同时，Unitprice不为0的数据
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

###### 筛选出来结果数量为0条，说明缺失Description值的数据，同时也不具备有效的UnitPrice值。

Description表示产品名称，UnitPrice表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

In [20]:
original_data[original_data['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

###### CustomerID表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些CustomerID缺失的交易数据仍然有效，因此保留变量为空的观察植。

#### **评估重复数据**

###### 根据数据变量的含义来看，虽然‘InvoiceNo’、‘StockCode’和‘CustomerID’都是唯一标识符，但一次交易可能包含多件商品，因此‘InvoiceNo’可以存在重复，不同交易也可以包含同件商品，一次‘StockCode’可以存在重复，顾客可以进行多次交易或下单多个商品，因此‘CustomerID’也可以存在重复。

#### **评估不一致数据**

###### 不一致数据可能存在于‘Country’变量中，需要查看是否哦存在多个不同值指代统一国家的情况。

In [26]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


###### 从Country变量值来看，“USA”、“United States”均表示在美国，“United Kingdom”、“UK”、“U.K."均在表示英国,因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

#### **评估无效或错误数据**

###### 可以通过DataFrame的describe方法，对数值统计信息进行快速了解。

In [30]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

###### 从输出结果来看，Quantity和UnitPrice存在负数，会对后续数值分析产生影响。

因此，我们筛选出Quantity数值为复数的观察值，进一步评估其含义。

In [32]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != "C") &(original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

###### 根据输出结果，猜想得到验证，当Quantity变量为负数时，观察植满足以下条件之一：
    ·当InvoiceNo以C开头，表示订单被取消
    ·UnitPrice为0，说明单价为0英镑
##### 这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将Quantity为负数的观察值剔除。

In [34]:
#筛选出UnitPrice为负数的观察值
original_data[original_data['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

###### 从输出结果看，UnitPrice为复数的观察值都是坏账调整，不属于实际商品交易数据，因此也在数据清理中将其删除。

## 清理数据

##### 根据前面评估部分得到的结论，需要进行的数据清理包括：
    · 把‘InvoiceDate’变量的数据类型转换为日期时间
    · 把‘Customer ID’的数据类型应为字符串
    · 把‘Description’变量缺失的观察值删除
    · 把‘Country’变量“USA”替换为“United States”，“UK”、“U.K."替换为“United Kingdom”
    · 把‘Quantity’变量值为负数的观察值删除
    · 把‘UnitPrice’变量值为负数的观察值删除除

In [38]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [39]:
# 把‘InvoiceDate’变量的数据类型转换为日期时间
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])

In [40]:
# 把‘Customer ID’的数据类型应为字符串
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [41]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].str.slice(0, -2)
cleaned_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

In [42]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   541909 non-null  object        
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 33.1+ MB


In [43]:
# 把‘Description’变量缺失的观察值删除，并查看删除后列空缺值个数
cleaned_data.dropna(subset = ["Description"], inplace = True)

In [44]:
cleaned_data["Description"].isnull().sum()

0

In [45]:
# 把‘Country’变量“USA”替换为“United States”，“UK”、“U.K."替换为“United Kingdom”
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"}, inplace = True)

In [46]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

In [47]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [48]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


In [49]:
# 把‘Quantity’变量值为负数的观察值删除
cleaned_data = cleaned_data[cleaned_data['Quantity'] >= 0]
len(cleaned_data[cleaned_data['Quantity'] < 0])

0

In [50]:
# 把‘UnitPrice’变量值为负数的观察值删除
cleaned_data = cleaned_data[cleaned_data['UnitPrice'] >= 0]
len(cleaned_data[cleaned_data['UnitPrice'] < 0])

0

## 保存清理后的数据

In [52]:
cleaned_data.to_csv("./data/e_commerce_clean.csv")